### PR TITLE
Sagantech changes

### PR DESCRIPTION
**Pull request by sagantech**
Patch browser changes, add checks and expose program name to host.

### CODE REVIEW

**Comment by baconpaul** [comment]
Hi @sagantech 

You can’t check in vst2 SDK to the repository.  For that reason alone this won’t get merged. You store it elsewhere then use the VST2SDK_DIR environment at build time. Sorry! Can you close this PR and put in one which doesn’t contain all 83 files.

As to the other changes, could you explain what they try and do in a commit message, comment, or issue? Like what do they do and how and why?

Thanks

**Comment by sagantech** [comment]
Hello Paul,

Thank you. Yes I mistakenly included the VST 2 SDK. I am a bit rusty on using github. I thought I did add commit messages. I am using VS 2017 so perhaps I missed something. I did a recap in the description of the pull request. So any help on this would be appreciated.

Thanks,

Jeremy Sagan
Sagan Technology
www.sagantech.biz

**Comment by baconpaul** [comment]
Hi!

Well the easiest thing to do is to make a new branch from master and bring your changes you need across. Store vst outside the surge tree and then you won’t check it in by accident. Then close this pr and send a request from the new branch

Generally commit descriptions have a short title and then a description. Think of that description as explaining this to folks who don’t know your diff. We’ve also had a habit of opening GitHub issues for things we change which lets folks pre comment on design

If you want some git hints there’s a git howto in the docs directory. I don’t know how to do things in visual studio.

Finally from looking at the diffs seems a few things are tabbed wrong. Do you have hard tabs on in vs? We are using spaces here. Also some things like a strncpy with a 63; where did 63 come from. A quick comment on things like that can help the next person along

Hope this helps and thanks for contributing. Our slack channel can often offer help on git and build things too if you use that.

Welcome!

**Comment by baconpaul** [comment]
Oh and if it helps, although it's a bit late I know, I finally got around to writing up the conventions @jsakkine, myself, and others have been using in pull request reviews and stuff.

https://github.com/surge-synthesizer/surge/blob/master/doc/developer.md

**Comment by baconpaul** [comment]
It still contains changes to vst3sdk which is a sub module we don’t own. If we have found bugs in vat3sdk we need to go through the Steinberg process to get them patched there then move our submodule pointer. That’s issue 373

I would be curious if in the 7 months since the submodule pointer was set any of these bugs got fixed over there

As to the diff in the surge code

1: the indents still look wrong. Hard tabs non aligned code and so on

2: I don’t know what the comment //js: means

3: I would like to restore this feature so would be happy to merge a clean pr with changes to surge which have a few comments are correctly aligned etc

4: I am very happy to have another developer working on this project!

I would love to move the submodule pointer.

File: src/common/SurgeSynthesizerIO.cpp [hunk 1]
```diff
@@ -200,6 +201,20 @@ void SurgeSynthesizer::loadRaw(const void* data, int size, bool preset)
    halt_engine = false;
    patch_loaded = true;
    refresh_editor = true;
+	if (patchid < 0)	//category.size() == 0)
+	{//JS: it is a setchunk so I look up the set current category and patch
```
Comment (jsakkine) [inline src/common/SurgeSynthesizerIO.cpp]: I don't understand this sentence.

File: src/common/SurgeSynthesizerIO.cpp [hunk 2]
```diff
@@ -200,6 +201,20 @@ void SurgeSynthesizer::loadRaw(const void* data, int size, bool preset)
...
+	    int cnt = storage.patch_list.size();
+		string name = storage.getPatch().name;
```
Comment (jsakkine) [inline src/common/SurgeSynthesizerIO.cpp]: Then indentation goes shifts suddenly 3 spaces right for no reason.

File: src/common/SurgeSynthesizerIO.cpp [hunk 3]
```diff
@@ -200,6 +201,20 @@ void SurgeSynthesizer::loadRaw(const void* data, int size, bool preset)
    halt_engine = false;
    patch_loaded = true;
    refresh_editor = true;
+	if (patchid < 0)	//category.size() == 0)
```
Comment (jsakkine) [inline src/common/SurgeSynthesizerIO.cpp]: Then indentation goes shifts suddenly 3 spaces right for no reason.

**Comment by jsakkine** [comment]
I postpone further review because without a proper description in the commit message it is impossible to reflect what you want to do is what your code change does. Only suggestion I have other than that, is to add something like `std::map<std::string, int> patchNameToID` to the class SurgeStorage.

**Comment by jsakkine** [comment]
@esaruoho, That's cool, but that does not have anything to do with the remarks about the commit message. Just did a PR about how commit messages ought to be written. Linking the issue is a good idea, but still the commit message also should describe the change, because there is infinite different (according to theory of computation) ways to solve any particular programming exercise. I want to be able to look at git log history and see what has been done without constantly jumping to Github. And also, how it was done.

File: src/common/SurgeSynthesizerIO.cpp (hunk 3 above)
Comment (baconpaul) [reply src/common/SurgeSynthesizerIO.cpp]: I think many of these are because the diff has hard tabs rather than spaces, also. I think.

**Comment by baconpaul** [comment]
> I postpone further review because without a proper description in the commit message it is impossible to reflect what you want to do is what your code change does. Only suggestion I have other than that, is to add something like `std::map<std::string, int> patchNameToID` to the class SurgeStorage.

Bigger problem is there is no reason to beleive patch name is 1:1 with patch. I can easily have a patch called “my pad” in 3 categories.

**Comment by jsakkine** [comment]
@baconpaul, Ugh, that is correct indeed! And it kind of proves why my nagging about describing the approach in the commit message is essential.  Otherwise, bad changes leak more easily. It kind of gives a "double check" both to the implementer (thought process after doing the change) and to the reviewer.

**Comment by baconpaul** [comment]
@sagantech oops I owe you a small apology. The diffs in GitHub are just from updating the vst3sdk pointer, not from you changing vst3sdk. See convo with @kzantow and I this morning (Sun 27) on slack.

So ignore my comments about "needing to talk to steinberg to merge those". I thought you had actually modified the files, not just pulled the lib.

But still, we have the problem that the code doesn't build with those updates. That's handled in issue #373. If you'd like to work with me on fixing those I could make a branch in surge with the update. Otherwise I will hack away at `baconpaul/vst3sdk-update-373` over the next day or two

But in any case, certainly can't merge this with that update and break all our builds! So perhaps taking this pull request just to your menu changes with the formatting and commenting approach is the way to go.

**Comment by jsakkine** [comment]
Please ping me when there is an update! Definitely want to re-review this.

**Comment by sagantech** [comment]
I don't understand the discussion here. It is true if you name the same patch in different categories it is possible when you reload a host document it will check the wrong category but I believe this can easily be fixed and is still 1. rare and 2. better behavior than what is currently happening. Why this code would not build makes no sense at all. I don't see any of the errors from the build related to my changes.

File: src/common/SurgeSynthesizerIO.cpp (hunk 1 above)
Comment (sagantech) [reply src/common/SurgeSynthesizerIO.cpp]: I changed the comment in the latest commit but basically setchunk is the method used by the host to send the plug-ins data to the plug-in to completely restore the saved patch.

File: src/common/SurgeSynthesizerIO.cpp (hunk 2 above)
Comment (sagantech) [reply src/common/SurgeSynthesizerIO.cpp]: I am not sure how to setup VS2017 to tab correctly for you guys.

**Comment by sagantech** [comment]
> Bigger problem is there is no reason to beleive patch name is 1:1 with patch. I can easily have a patch called “my pad” in 3 categories.

This problem has been addressed in the latest commit.

**Comment by baconpaul** [comment]
It doesn’t build because https://github.com/surge-synthesizer/surge/commit/c9b818b9621d3c85aa4784b500ec2b76bb27b9aa updates the vst sdk pointer and as discussed in #373 that requires code changes. The automated builds will show you the output.

**Comment by sagantech** [comment]
It is a shame that that is attributed to me. I just did what Keith told me to do.

Jeremy Sagan
Sagan Technology
www.sagantech.biz

**Comment by baconpaul** [comment]
I wasn't ascribing blame in a pejorative sense. Just explaining what was wrong with the PR!

You need to do a rebase to knock that commit out basically and force push.  Probably smart to at the same time squash all the other commits into one with a single commit message. Have you ever done a git rebase? If not, I'll update the git doc with a section on how to do it and we can give it a whirl.

**Comment by baconpaul** [comment]
Oh and I'm well along the way to getting #373 working. So this may be less of an issue soon!

**Comment by baconpaul** [comment]
>  It is true if you name the same patch in different categories it is possible when you reload a host document it will check the wrong category but I believe this can easily be fixed and is still 1. rare and 2. better behavior than what is currently happening. 

My comment was directed at @jsakkine suggestion to use a std::map which wouldn't work. Your test in the loop of category equivalence looks correct to me, although I haven't looked at your code that closely.

**Comment by sagantech** [comment]
No, I am a total gitNuBe

Jeremy Sagan
Sagan Technology
www.sagantech.biz

**Comment by sagantech** [comment]
I tested it by putting the same named patch in two categories, saving a document with the second of the same name and then reloading. I then repeated by saving in the first category and reloaded. In both cases the checks in the category  and the patch were correct. I then removed the second duplicate named patch and reloaded the first document and no checks were present. This is what I would expect. Of course the checkmark on the category requires an update to VSTGUI but It is very small, simple and I think it is beneficial to that repository.

Jeremy Sagan
Sagan Technology
www.sagantech.biz

**Comment by baconpaul** [comment]
Yeah the problem is that the small update to vst3sdk also puts everything into a namespace. It's tedious to fix but I'm working on it.

**Comment by sagantech** [comment]
Yes, I did all of that work too for my 1.6b build before I connected here. Very tedious namespace VSTGUI { stuff }

**Comment by baconpaul** [comment]
OK @sagantech here's what you want to do

In your repo with the code checked out at your latest version type

```
git rebase -i master
```

this will bring up a screen with each of your commits with the word 'pick' next to it

Next to the one which moves the vst3sdk pointer c9b818b change the word 'pick' to 'drop'

In all the rest of them except the first change the word 'pick' to 'squash'. So it will look like pick drop squash squash squash or whatever. Write that file

You then get an opportunity to write a new commit message. Write that using the outline that's in the developer doc. You get the idea. Then save that

This will change your git branch to being 'as if' it only had one commit and didn't include the vst3sdk pointer move.

Then do a 'git update origin --force' and this issue will have one commit and a non-breaking change.

**Comment by sagantech** [comment]
"Next to the one which moves the vst3sdk pointer c9b818b change the word 'pick' to 'drop'"

Stuck on this step. In cmd.exe don't see how to do that.

File: src/common/gui/CPatchBrowser.cpp [hunk 4]
```diff
@@ -127,6 +127,8 @@ void CPatchBrowser::populatePatchMenuForCategory( int c, COptionMenu *contextMen
             auto actionItem = new CCommandMenuItem(name);
             auto action = [this, p](CCommandMenuItem* item) { this->loadPatch(p); };
             
+			if (p == current_patch)
```
Comment (baconpaul) [inline src/common/gui/CPatchBrowser.cpp]: The indentation here still looks like it has hard tabs?

File: src/common/gui/CPatchBrowser.cpp [hunk 5]
```diff
@@ -155,8 +157,10 @@ void CPatchBrowser::populatePatchMenuForCategory( int c, COptionMenu *contextMen
         
         if (!single_category)
         {
-            contextMenu->addEntry(subMenu, name);
-            subMenu->forget(); // Important, so that the refcounter gets it right
+            CMenuItem *entry = contextMenu->addEntry(subMenu, name);
+      		if (c == current_category)
+				entry->setChecked(true);
+           subMenu->forget(); // Important, so that the refcounter gets it right
```
Comment (baconpaul) [inline src/common/gui/CPatchBrowser.cpp]: The indentation here is off by one space in on the subMenu and looks like it has hard tabs on the if.

File: src/vst2/Vst2PluginInstance.h [hunk 6]
```diff
@@ -28,6 +28,7 @@ class Vst2PluginInstance : public AudioEffectX
    virtual VstInt32 processEvents(VstEvents* ev);
 
    // Host->Client calls
+   virtual bool Vst2PluginInstance::getProgramNameIndexed(VstInt32 category, VstInt32 index, char* text);
```
Comment (baconpaul) [inline src/vst2/Vst2PluginInstance.h]: While it's valid to fully qualify the name here (Vst2PluginIntance::getProgramNameIndexed) you don't need to and we generally don't I think you can remove the qualifier.

File: src/vst2/Vst2PluginInstance.cpp [hunk 7]
```diff
@@ -294,9 +294,24 @@ void Vst2PluginInstance::setProgramName(char* name)
...
+//   strcpy(name, programName);
+//}
 void Vst2PluginInstance::getProgramName(char* name)
 {
-   strcpy(name, programName);
+//   strcpy(name, programName);
+	getProgramNameIndexed(0, 0, name);
+}
+bool Vst2PluginInstance::getProgramNameIndexed (VstInt32 category, VstInt32 index, char* text)
+{//JS:
```
Comment (baconpaul) [inline src/vst2/Vst2PluginInstance.cpp]: I think JS indicates your initials, but we don't need that; git blame will show us you wrote it. Also JS is (funnily) not a unique initial set among the 3 people who have made C++ PRs in the last 2 weeks :)

**Comment by baconpaul** [comment]
Oh I hate to tell you this. Your latest commit reintroduced your vst3sdk change so you have to rebase again :(

**Comment by baconpaul** [comment]
Tell you what, @sagantech - this change will sit much more easily after I move the vst3sdk pointer. Let’s do this. When I have that move done I will pull and merge your commit from your branch (which will keep you having credit for it in the log and everything) and make a fresh post-pointer-move PR.

You’ll jump out a window unpicking the commit you did above. So if it’s OK with you lets pause this for a few days and get to it when the pointer is moved and we are all back at master.

If you really want to unpick the commit in git, go ahead of course! But I can imagine you don’t want to... and it’s easy for me to do so once I have the pointer moved.

**Comment by baconpaul** [comment]
Closing this one; new one coming soon including these diffs.

**Pull request closed without merging**